File: fly/src/main/java/com/tiamosu/fly/integration/gson/data/JSONArrayTypeAdapter.java
package com.tiamosu.fly.integration.gson.data;

import com.google.gson.JsonElement;
import com.google.gson.TypeAdapter;
import com.google.gson.internal.bind.TypeAdapters;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.IOException;

/**
 * author : Android 轮子哥
 * github : https://github.com/getActivity/GsonFactory
 * time   : 2021/09/30
 * desc   : JSONObject 类型解析适配器
 */
public class JSONArrayTypeAdapter extends TypeAdapter<JSONArray> {
    private final TypeAdapter<JsonElement> mProxy = TypeAdapters.JSON_ELEMENT;

    @Override
    public JSONArray read(JsonReader in) throws IOException {
        final JsonElement read = mProxy.read(in);
        if (read.isJsonArray()) {
            try {
                return new JSONArray(read.toString());
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return new JSONArray();
    }

    @Override
    public void write(JsonWriter out, JSONArray value) throws IOException {
        if (value == null) {
            out.nullValue();
            return;
        }
        mProxy.write(out, mProxy.fromJson(value.toString()));
    }
}

File: fly/src/main/java/com/tiamosu/fly/integration/gson/data/JSONArrayTypeAdapter.kt
package com.tiamosu.fly.integration.gson.data

import com.google.gson.JsonElement
import com.google.gson.TypeAdapter
import com.google.gson.internal.bind.TypeAdapters
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import org.json.JSONArray

/**
 * @author tiamosu
 * @date 2022/7/3
 *
 * 描述：JSONArray 类型解析适配器
 */
class JSONArrayTypeAdapter : TypeAdapter<JSONArray>() {
    private val proxy: TypeAdapter<JsonElement> by lazy { TypeAdapters.JSON_ELEMENT }

    override fun read(`in`: JsonReader?): JSONArray {
        if (`in` == null) {
            return JSONArray()
        }
        val read = proxy.read(`in`)
        if (read.isJsonArray) {
            kotlin.runCatching {
                return JSONArray(read.toString())
            }
        }
        return JSONArray()
    }

    override fun write(out: JsonWriter?, value: JSONArray?) {
        if (value == null) {
            out?.nullValue()
            return
        }
        proxy.write(out, proxy.fromJson(value.toString()))
    }
}